// is/dist/index.js
/* MAIN */
import isArguments from './is_arguments.js';
import isArray from './is_array.js';
import isArrayBuffer from './is_array_buffer.js';
import isArrayLike from './is_array_like.js';
import isArrayLikeObject from './is_array_like_object.js';
import isArrowFunction from './is_arrow_function.js';
import isAsyncFunction from './is_async_function.js';
import isAsyncGeneratorFunction from './is_async_generator_function.js';
import isAsyncIterable from './is_async_iterable.js';
import isAttribute from './is_attribute.js';
import isBigInt from './is_bigint.js';
import isBigInt64Array from './is_bigint64array.js';
import isBigUint64Array from './is_biguint64array.js';
import isBlob from './is_blob.js';
import isBoolean from './is_boolean.js';
import isBoundFunction from './is_bound_function.js';
import isBoxedBigInt from './is_boxed_bigint.js';
import isBoxedBoolean from './is_boxed_boolean.js';
import isBoxedNumber from './is_boxed_number.js';
import isBoxedPrimitive from './is_boxed_primitive.js';
import isBoxedString from './is_boxed_string.js';
import isBoxedSymbol from './is_boxed_symbol.js';
import isBuffer from './is_buffer.js';
import isClass from './is_class.js';
import isComment from './is_comment.js';
import isDataView from './is_data_view.js';
import isDate from './is_date.js';
import isDocument from './is_document.js';
import isDocumentFragment from './is_document_fragment.js';
import isDocumentType from './is_document_type.js';
import isElement from './is_element.js';
import isEmpty from './is_empty.js';
import isError from './is_error.js';
import isEven from './is_even.js';
import isFalsy from './is_falsy.js';
import isFinite from './is_finite.js';
import isFloat from './is_float.js';
import isFloat32Array from './is_float32array.js';
import isFloat64Array from './is_float64array.js';
import isFrozen from './is_frozen.js';
import isFunction from './is_function.js';
import isGeneratorFunction from './is_generator_function.js';
import isInt16Array from './is_int16array.js';
import isInt32Array from './is_int32array.js';
import isInt8Array from './is_int8array.js';
import isInteger from './is_integer.js';
import isIterable from './is_iterable.js';
import isLength from './is_length.js';
import isMap from './is_map.js';
import isNaN from './is_nan.js';
import isNative from './is_native.js';
import isNegativeZero from './is_negative_zero.js';
import isNil from './is_nil.js';
import isNode from './is_node.js';
import isNull from './is_null.js';
import isNumber from './is_number.js';
import isNumberLike from './is_number_like.js';
import isObject from './is_object.js';
import isObjectLike from './is_object_like.js';
import isOdd from './is_odd.js';
import isPlainObject from './is_plain_object.js';
import isPrimitive from './is_primitive.js';
import isPromise from './is_promise.js';
import isPrototype from './is_prototype.js';
import isRegExp from './is_regexp.js';
import isSealed from './is_sealed.js';
import isSafeInteger from './is_safe_integer.js';
import isSet from './is_set.js';
import isSharedArrayBuffer from './is_shared_array_buffer.js';
import isString from './is_string.js';
import isSymbol from './is_symbol.js';
import isText from './is_text.js';
import isTruthy from './is_truthy.js';
import isTypedArray from './is_typed_array.js';
import isUint16Array from './is_uint16array.js';
import isUint32Array from './is_uint32array.js';
import isUint8Array from './is_uint8array.js';
import isUint8ClampedArray from './is_uint8clampedarray.js';
import isUndefined from './is_undefined.js';
import isWeakMap from './is_weak_map.js';
import isWeakRef from './is_weak_ref.js';
import isWeakReferable from './is_weak_referable.js';
import isWeakSet from './is_weak_set.js';
import isWindow from './is_window.js';
/* EXPORT */
export { isArguments, isArray, isArrayBuffer, isArrayLike, isArrayLikeObject, isArrowFunction, isAsyncFunction, isAsyncGeneratorFunction, isAsyncIterable, isAttribute, isBigInt, isBigInt64Array, isBigUint64Array, isBlob, isBoolean, isBoundFunction, isBoxedBigInt, isBoxedBoolean, isBoxedNumber, isBoxedPrimitive, isBoxedString, isBoxedSymbol, isBuffer, isClass, isComment, isDataView, isDate, isDocument, isDocumentFragment, isDocumentType, isElement, isEmpty, isError, isEven, isFalsy, isFinite, isFloat, isFloat32Array, isFloat64Array, isFrozen, isFunction, isGeneratorFunction, isInt16Array, isInt32Array, isInt8Array, isInteger, isIterable, isLength, isMap, isNaN, isNative, isNegativeZero, isNil, isNode, isNull, isNumber, isNumberLike, isObject, isObjectLike, isOdd, isPlainObject, isPrimitive, isPromise, isPrototype, isRegExp, isSealed, isSafeInteger, isSet, isSharedArrayBuffer, isString, isSymbol, isText, isTruthy, isTypedArray, isUint16Array, isUint32Array, isUint8Array, isUint8ClampedArray, isUndefined, isWeakMap, isWeakRef, isWeakReferable, isWeakSet, isWindow };
