// configuration/dist/providers/abstract_file.js
/* IMPORT */
import ProviderMemory from './memory.js';
/* MAIN */
class ProviderAbstractFile extends ProviderMemory {
    /* CONSTRUCTOR */
    constructor(options) {
        super(options);
        this.watching = !!options.watch;
        this.writeOptions = options.writeOptions;
        this.writeSyncOptions = options.writeSyncOptions;
        this.swap(options.path, true);
    }
    /* API */
    dispose() {
        this.unwatch();
    }
    swap(path, _initial = false) {
        if (path === this.path)
            return;
        this.dispose();
        this.path = path;
        this.init();
        if (!_initial)
            this.triggerChange();
        if (this.watching)
            this.watch();
    }
    watch() {
        if (!this.path)
            return;
        const path = this.path;
        this.watcher = this.fileWatch(path, async () => {
            const { dataRaw } = await this.read();
            if (path !== this.path)
                return;
            if (this.isEqual(dataRaw))
                return;
            super.write(dataRaw, true);
        });
    }
    unwatch() {
        if (!this.watcher)
            return;
        this.watcher.close();
        delete this.watcher;
    }
}
/* EXPORT */
export default ProviderAbstractFile;
