// configuration/dist/providers/memory.js
/* IMPORT */
import cloneDeep from 'plain-object-clone';
import PathProp from '../utils/pp.js';
import Type from '../utils/type.js';
import ProviderAbstract from './abstract.js';
/* MAIN */
class ProviderMemory extends ProviderAbstract {
    /* API */
    async read() {
        return this.readSync();
    }
    readSync() {
        const data = this.data ?? cloneDeep(this.defaults);
        const dataRaw = this.dataRaw ?? this.defaultsRaw;
        return { data, dataRaw };
    }
    async write(data, force = false) {
        return this.writeSync(data, force);
    }
    writeSync(data, force = false) {
        if (!force && this.isEqual(data))
            return;
        if (Type.isString(data)) {
            this.data = PathProp.unflat(this.dataParser.parse(data) ?? this.defaults);
            this.dataRaw = data;
            this.dataSchema = this.filterer(this.data);
        }
        else {
            this.data = PathProp.unflat(data);
            this.dataRaw = this.dataParser.stringify(data, this.dataRaw) ?? this.defaultsRaw;
            this.dataSchema = this.filterer(this.data);
        }
        this.triggerChange();
    }
}
/* EXPORT */
export default ProviderMemory;
